File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.py
import sys
input = sys.stdin.readline

s, p = map(int, input().split())
dna = input().rstrip()
a, c, g, t = map(int, input().split())

dict_part = {'A': 0,'C': 0,'G': 0,'T': 0}
start, end = 0, p-1

cnt = 0

word = dna[start:end]

for w in word:
    dict_part[w] += 1

while end < s:
    # for w in word:
    #     dict_part[w] += 1
    dict_part[dna[end]] += 1
    if dict_part['A'] >= a and dict_part['C'] >= c and dict_part['G'] >= g and dict_part['T'] >= t:
        cnt += 1
    dict_part[dna[start]] -= 1
    start += 1
    end += 1


print(cnt)

File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        int s = Integer.parseInt(st.nextToken());
        int p = Integer.parseInt(st.nextToken());
        
        String dna = br.readLine().trim();
        
        st = new StringTokenizer(br.readLine());
        int a = Integer.parseInt(st.nextToken());
        int c = Integer.parseInt(st.nextToken());
        int g = Integer.parseInt(st.nextToken());
        int t = Integer.parseInt(st.nextToken());
        
        Map<Character, Integer> dictPart = new HashMap<>();
        dictPart.put('A', 0);
        dictPart.put('C', 0);
        dictPart.put('G', 0);
        dictPart.put('T', 0);
        
        int start = 0, end = p - 1;
        int cnt = 0;
        String word = dna.substring(start, end);

        for (char w : word.toCharArray()) {
            dictPart.put(w, dictPart.get(w) + 1);
        }

        while (end < s) {
            dictPart.put(dna.charAt(end), dictPart.get(dna.charAt(end)) + 1);
            
            if (dictPart.get('A') >= a && dictPart.get('C') >= c &&
                dictPart.get('G') >= g && dictPart.get('T') >= t) {
                cnt++;
            }

            dictPart.put(dna.charAt(start), dictPart.get(dna.charAt(start)) - 1);
            start++;
            end++;
        }

        System.out.println(cnt);
    }
}